File: medical_transcriber/ios/Runner/NativeAudioPlugin.swift
import Flutter
import AVFoundation

class NativeAudioPlugin: NSObject, FlutterPlugin, FlutterStreamHandler {

    static let shared = NativeAudioPlugin()

    private var eventSink: FlutterEventSink?
    private var routeSink: FlutterEventSink?
    private var bufferSink: FlutterEventSink?

    static func register(with registrar: FlutterPluginRegistrar) {
        let instance = NativeAudioPlugin()

        let methodChannel = FlutterMethodChannel(name: "native_audio_channel", binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(instance, channel: methodChannel)

        let eventChannel = FlutterEventChannel(name: "native_audio_levels", binaryMessenger: registrar.messenger())
        eventChannel.setStreamHandler(instance)

        let routeChannel = FlutterEventChannel(name: "native_audio_routes", binaryMessenger: registrar.messenger())
        routeChannel.setStreamHandler(instance)

        let bufferChannel = FlutterEventChannel(name: "native_audio_buffers", binaryMessenger: registrar.messenger())
        bufferChannel.setStreamHandler(instance)
    }

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "startRecording":
            if let args = call.arguments as? [String: Any],
            let sessionId = args["sessionId"] as? String {
                try? NativeAudioManager.shared.startRecording(sessionId: sessionId)
                result(nil)
            } else {
                result(FlutterError(code: "INVALID_ARGUMENTS", message: "Missing sessionId", details: nil))
            }
        case "pauseRecording":
            NativeAudioManager.shared.pauseRecording()
            result(nil)
        case "resumeRecording":
            NativeAudioManager.shared.resumeRecording()
            result(nil)
        case "stopRecording":
            if let args = call.arguments as? [String: Any],
            let isLast = args["isLast"] as? Bool {
                NativeAudioManager.shared.stopRecording(isLast: isLast)
            }
            result(nil)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    func emitAudioLevel(level: Float) {
        levelSink?(level)
    }

    func emitChunkEvent(fileUrl: URL, chunkNumber: Int, isLast: Bool) {
        let event: [String: Any] = [
            "filePath": fileUrl.absoluteString,
            "chunkNumber": chunkNumber,
            "isLast": isLast
        ]
        bufferSink?(event)
    }
}

class AudioLevelStreamHandler: NSObject, FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        NativeAudioPlugin.audioLevelSink = events
        return nil
    }
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        NativeAudioPlugin.audioLevelSink = nil
        return nil
    }
}

class AudioChunkStreamHandler: NSObject, FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        NativeAudioPlugin.chunkSink = events
        return nil
    }
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        NativeAudioPlugin.chunkSink = nil
        return nil
    }
}

class AudioRouteStreamHandler: NSObject, FlutterStreamHandler {
    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        NativeAudioPlugin.routeSink = events
        return nil
    }
    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        NativeAudioPlugin.routeSink = nil
        return nil
    }
}

File: medical_transcriber/android/app/src/main/kotlin/com/example/medical_transcriber/NativeAudioPlugin.kt
package com.example.medical_transcriber

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.BinaryMessenger
import java.io.File
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch

class NativeAudioPlugin(
    private val context: Context,
    messenger: BinaryMessenger
) : MethodChannel.MethodCallHandler {

    companion object {
        lateinit var instance: NativeAudioPlugin
        lateinit var shared: NativeAudioPlugin
    }

    private val methodChannel = MethodChannel(messenger, "native_audio_channel")
    private val levelChannel = EventChannel(messenger, "native_audio_levels")
    private val routeChannel = EventChannel(messenger, "native_audio_routes")
    private val bufferChannel = EventChannel(messenger, "native_audio_buffers")

    private var levelSink: EventChannel.EventSink? = null
    private var routeSink: EventChannel.EventSink? = null
    private var bufferSink: EventChannel.EventSink? = null

    private val pluginScope = CoroutineScope(Dispatchers.IO + SupervisorJob())

    private var recorder: ChunkAudioRecorder? = null

    init {
        instance = this

        methodChannel.setMethodCallHandler(this)

        levelChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                levelSink = events
            }

            override fun onCancel(arguments: Any?) {
                levelSink = null
            }
        })

        routeChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                routeSink = events
            }

            override fun onCancel(arguments: Any?) {
                routeSink = null
            }
        })

        bufferChannel.setStreamHandler(object : EventChannel.StreamHandler {
            override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
                bufferSink = events
            }

            override fun onCancel(arguments: Any?) {
                bufferSink = null
            }
        })
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "startRecording" -> {
                val sessionId = call.argument<String>("sessionId")!!
                recorder = ChunkAudioRecorder(context, sessionId)
                recorder?.start()
                result.success(null)
            }

            "pauseRecording" -> {
                recorder?.pause()
                result.success(null)
            }

            "resumeRecording" -> {
                recorder?.resume()
                result.success(null)
            }

            "stopRecording" -> {
                val isLast = call.argument<Boolean>("isLast") ?: true
                pluginScope.launch {
                    recorder?.stop(isLast)
                    result.success(null)
                }
            }

            else -> result.notImplemented()
        }
    }

    fun emitAudioLevel(level: Double) {
        Handler(Looper.getMainLooper()).post {
            levelSink?.success(level)
        }
    }


    fun emitRouteChange(route: String) {
        Handler(Looper.getMainLooper()).post {
            routeSink?.success(route)
        }
    }

    fun emitChunkEvent(payload: Map<String, Any>) {
        Handler(Looper.getMainLooper()).post {
            bufferSink?.success(payload)
        }
    }
}
